### DISCUSSION TITLE
Testing on a huge graph.

### DISCUSSION THREAD

**Discussion by JiaruiWang** [pyg-team/pytorch_geometric#4538] [Q&A]
My task is a node classification problem on a 60M nodes graph. 20M nodes are labeled, and 40M nodes are unlabeled. I created the dataset with a 16M nodes training mask, a 2M nodes validation mask, and a 2M nodes testing mask, out of 20M labeled nodes.
I feed the 16M training nodes into [NeighborLoader](https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html#torch_geometric.loader.NeighborLoader) to generate the data loader for training. Sampled subgraphs can fit into the GPU memory.

In the evaluation process, all the examples pass the whole graph into the model, then get the output[test_mask]. This will cause the GPU OOM. If compute the evaluation process on the CPU, it is too slow. 

What's the best practice for it? 

Thank you very much

**Comment by rusty1s** [comment]
You are right that one cannot do full-batch inference on CPU on such a giant graph. The alternative is to also use neighbor sampling during inference. To reduce the variance due to sampling/dropout of edges, you can either try to sample all neighbors in the inference loader (`num_neighbors=[-1] * num_layers`), or increase the number of neighbors. You can also re-run inference multiple times and perform ensembling.
